DUE 2021-09-21: Packages including omp.h after Rincludes.h

That is

AMORE BNN BRISC OpenMPController RFCCA RFpredInterval RLT
RandomFields RandomFieldsUtils RandomForestsGLS RhpcBLASctl RxODE
STB TMB brnn coop data.table excursions float glinternet glmmTMB
gmGeostats graphscan hashr hutilscpp influenceR iq kit likeLTD
mRMRe magick meanShiftR meanr mnlogit poismf **poppr** praznik rFerns
randomForestSRC rasterKernelEstimates rehh simputation spBayes
spNNGP stringdist sylcount unfoldr


Compilation logs can be seen at https://www.stats.ox.ac.uk/pub/bdr/clang13/ but beware these may only show the first error.

<details>
<summary>Compiliation Log for poppr</summary>

```
* installing *source* package ‘poppr’ ...
** package ‘poppr’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
make[1]: Entering directory '/data/gannet/ripley/R/packages/tests-clang13/poppr/src'
/usr/local/clang13/bin/clang -I"/data/gannet/ripley/R/R-clang-trunk/include" -DNDEBUG   -isystem /usr/local/clang13/include  -fopenmp -fpic  -g -O3 -Wall -pedantic  -c adjust_missing.c -o adjust_missing.o
/usr/local/clang13/bin/clang -I"/data/gannet/ripley/R/R-clang-trunk/include" -DNDEBUG   -isystem /usr/local/clang13/include  -fopenmp -fpic  -g -O3 -Wall -pedantic  -c bitwise_distance.c -o bitwise_distance.o
In file included from bitwise_distance.c:50:
/usr/local/clang13/lib/clang/13.0.0/include/omp.h:483:39: error: expected 'match' clause on 'omp declare variant' directive
    #pragma omp begin declare variant match(device={kind(host)})
                                      ^
/data/gannet/ripley/R/R-clang-trunk/include/Rinternals.h:967:17: note: expanded from macro 'match'
#define match                   Rf_match
                                ^
In file included from bitwise_distance.c:50:
/usr/local/clang13/lib/clang/13.0.0/include/omp.h:486:39: error: expected 'match' clause on 'omp declare variant' directive
    #pragma omp begin declare variant match(device={kind(nohost)})
                                      ^
/data/gannet/ripley/R/R-clang-trunk/include/Rinternals.h:967:17: note: expanded from macro 'match'
#define match                   Rf_match
                                ^
bitwise_distance.c:815:7: warning: variable 'chr_length' set but not used [-Wunused-but-set-variable]
  int chr_length;
      ^
bitwise_distance.c:1190:7: warning: variable 'chr_length' set but not used [-Wunused-but-set-variable]
  int chr_length;
      ^
2 warnings and 2 errors generated.
make[1]: *** [/data/gannet/ripley/R/R-clang-trunk/etc/Makeconf:168: bitwise_distance.o] Error 1
make[1]: Leaving directory '/data/gannet/ripley/R/packages/tests-clang13/poppr/src'
ERROR: compilation failed for package ‘poppr’
* removing ‘/data/gannet/ripley/R/packages/tests-clang13/Libs/poppr-lib/poppr’
Command exited with non-zero status 1
Time 0:01.95, 0.75 + 0.17
```

</details>

The imminent clang 13.0.0 has an omp.h including

```c
    /* OpenMP 5.1 */
...

#   if defined(_OPENMP) && _OPENMP >= 201811
    #pragma omp begin declare variant match(device={kind(host)})
    static inline int omp_is_initial_device(void) { return 1; }
    #pragma omp end declare variant
    #pragma omp begin declare variant match(device={kind(nohost)})
    static inline int omp_is_initial_device(void) { return 0; }
    #pragma omp end declare variant
#   endif
```

whose use of match causes a compilation error if included after Rinternals.h.  'Writing R Extensions' presents various workarounds, the main one being the strong advice to include system headers before R headers.

Please correct before 2021-09-21 to safely retain your package on CRAN.

-- 
Brian D. Ripley,                  ripley@stats.ox.ac.uk
Emeritus Professor of Applied Statistics, University of Oxford

--- a/CRAN-RELEASE
+++ b/CRAN-RELEASE
@@ -1,2 +1,2 @@
-This package was submitted to CRAN on 2021-05-22.
-Once it is accepted, delete this file and tag the release (commit f0601677).
+This package was submitted to CRAN on 2021-09-06.
+Once it is accepted, delete this file and tag the release (commit 4013eba7).

--- a/DESCRIPTION
+++ b/DESCRIPTION
@@ -1,7 +1,7 @@
 Package: poppr
 Type: Package
 Title: Genetic Analysis of Populations with Mixed Reproduction
-Version: 2.9.2
+Version: 2.9.3
 Authors@R: c(person(c("Zhian", "N."), "Kamvar", role = c("cre", "aut"),
     email = "zkamvar@gmail.com", comment = c(ORCID = "0000-0003-1458-7108")),
     person(c("Javier", "F."), "Tabima", role = "aut",

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,3 +1,12 @@
+poppr 2.9.3
+===========
+
+CRAN MAINTENANCE
+----------------
+
+* C headers were re-arranged to place R-specific headers _after_ OMP headers to
+  avoid problems in clang 13 (@zkamvar, https://github.com/grunwaldlab/poppr/pull/246).
+
 poppr 2.9.2
 ===========
 

--- a/cran-comments.md
+++ b/cran-comments.md
@@ -1,11 +1,11 @@
-# Poppr version 2.9.2
+# Poppr version 2.9.3
 
-This update fixes a test failure pointed out by CRAN on 2021-05-21
+This update rearranges the C files to declare all non-R headers before R headers to address the openmp failures in clang13 as pointed out by Prof. Ripley. 
 
 ## Test environments
 
-* local macOS install, R 4.1.0
-* local ubuntu 20.04 install, R 4.1.0
+* local macOS install, R 4.1.1
+* local ubuntu 20.04 install, R 4.1.1, R devel
 * windows R Under development (unstable) 
 
 ## R CMD check results

--- a/src/bitwise_distance.c
+++ b/src/bitwise_distance.c
@@ -36,19 +36,19 @@
 
 #include <stdio.h>
 #include <stdint.h>
-#include <Rinternals.h>
-#include <R_ext/Utils.h>
-#include <Rdefines.h>
-#include <R.h>
 #include <math.h>
 #include <time.h>
 #include <string.h>
 #include <stdlib.h>
-
 // Include openMP if the compiler supports it
 #ifdef _OPENMP
 #include <omp.h>
 #endif
+#include <Rinternals.h>
+#include <R_ext/Utils.h>
+#include <Rdefines.h>
+#include <R.h>
+
 
 // Assumptions:
 //  All genotypes have the same number of SNPs available.

--- a/src/init.c
+++ b/src/init.c
@@ -1,6 +1,6 @@
+#include <stdlib.h> // for NULL
 #include <R.h>
 #include <Rinternals.h>
-#include <stdlib.h> // for NULL
 #include <R_ext/Rdynload.h>
 
 /* FIXME: 

--- a/src/mlg_clustering.c
+++ b/src/mlg_clustering.c
@@ -35,14 +35,14 @@
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
 
 #include <stdio.h>
-#include <Rinternals.h>
-#include <R_ext/Utils.h>
-#include <Rdefines.h>
-#include <R.h>
 #include <math.h>
 #include <time.h>
 #include <string.h>
 #include <stdlib.h>
+#include <Rinternals.h>
+#include <R_ext/Utils.h>
+#include <Rdefines.h>
+#include <R.h>
 
 // Thu Apr 13 08:42:12 2017 ------------------------------
 // This code produces bugs when run on Fedora with multiple threads. Because of

--- a/src/msn.c
+++ b/src/msn.c
@@ -35,10 +35,6 @@
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
 
 #include <stdio.h>
-#include <Rinternals.h>
-#include <R_ext/Utils.h>
-#include <Rdefines.h>
-#include <R.h>
 #include <math.h>
 #include <time.h>
 #include <string.h>
@@ -47,6 +43,10 @@
 #ifdef _OPENMP
 #include <omp.h>
 #endif
+#include <Rinternals.h>
+#include <R_ext/Utils.h>
+#include <Rdefines.h>
+#include <R.h>
 
 
 /*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

--- a/src/poppr_distance.c
+++ b/src/poppr_distance.c
@@ -34,14 +34,14 @@
 #
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
 #include <stdio.h>
-#include <Rinternals.h>
-#include <Rdefines.h>
-#include <R.h>
-#include <R_ext/Utils.h>
 #include <math.h>
 #include <time.h>
 #include <string.h>
 #include <stdlib.h>
+#include <Rinternals.h>
+#include <Rdefines.h>
+#include <R.h>
+#include <R_ext/Utils.h>
 int perm_count;
 
 SEXP pairwise_covar(SEXP pair_vec);

